WordPress/openverse-catalog
# Wikimedia: re-attempt large batches with reduced parameter selection

## Fixes




Fixes #725 by @stacimc

## Description




**Author's note**: This...is admittedly a really massive and complicated PR :sweat_smile: _(at least it seems that way to me!)_ Wikimedia's API is incredibly complicated, their documentation is hard to navigate, and there's a lot of "magic" that happens with it that can make it hard to introspect & grok. I've tried to summarize this all as best as I could in the DAG documentation, hopefully it's not too loquacious.

This PR changes the behavior of the Wikimedia ingestion to detect when a particular batch has been iterated over more than a certain number of times, and adjust the query parameters for the next request to try and reduce the amount of data returned. It does this by _excluding_ the fields which were returning massive amounts of data. As of today, none of those fields are strictly _necessary_, merely useful. Since the items that would cause the issue in question would rarely complete during the task's iteration anyway, having them with _partial_ data seems much better than not having them at all.

Since we're storing more state on the ingester instance as well, I've modified the tests so each test gets a fresh instantiation. I also confirmed that modifying values _on an instance_ doesn't affect the values that are given for new instances of the class (e.g. changing the max iteration cutoff for an instance doesn't change it at the class level).

## Testing Instructions



**Author's note**: this is hairy and complicated and I'm sorry!!

1. `just recreate`
2. Change the following (just for testing purposes, example diff shown below)
    1. Set `max_page_iteration_before_give_up` to 10
    2. Change the start date of Wikimedia to `2023-02-01`
    3. Change the DAG factory to set all dated DAGs' `catchup` to `False`

```diff
diff --git a/openverse_catalog/dags/providers/provider_api_scripts/wikimedia_commons.py b/openverse_catalog/dags/providers/provider_api_scripts/wikimedia_commons.py
index 08faa95..91cc462 100644
--- a/openverse_catalog/dags/providers/provider_api_scripts/wikimedia_commons.py
+++ b/openverse_catalog/dags/providers/provider_api_scripts/wikimedia_commons.py
@@ -132,7 +132,7 @@ class WikimediaCommonsDataIngester(ProviderDataIngester):
     # These can be very large, so we need to limit the number of attempts otherwise
     # the DAG will hit the timeout when it comes across these items.
     # See: https://github.com/WordPress/openverse-catalog/issues/725
-    max_page_iteration_before_give_up = 100
+    max_page_iteration_before_give_up = 10
 
     image_mediatypes = {"BITMAP", "DRAWING"}
     audio_mediatypes = {"AUDIO"}
diff --git a/openverse_catalog/dags/providers/provider_dag_factory.py b/openverse_catalog/dags/providers/provider_dag_factory.py
index c460223..e00fa87 100644
--- a/openverse_catalog/dags/providers/provider_dag_factory.py
+++ b/openverse_catalog/dags/providers/provider_dag_factory.py
@@ -305,7 +305,7 @@ def create_provider_api_workflow_dag(conf: ProviderWorkflow):
         max_active_runs=conf.max_active_runs,
         start_date=conf.start_date,
         schedule=conf.schedule_string,
-        catchup=conf.dated,  # catchup is turned on for dated DAGs to allow backfilling
+        catchup=False,  # conf.dated,  # catchup is turned on for dated DAGs to allow backfilling
         doc_md=conf.doc_md,
         tags=[
             "provider",
diff --git a/openverse_catalog/dags/providers/provider_workflows.py b/openverse_catalog/dags/providers/provider_workflows.py
index c0d699f..9556c8a 100644
--- a/openverse_catalog/dags/providers/provider_workflows.py
+++ b/openverse_catalog/dags/providers/provider_workflows.py
@@ -202,7 +202,7 @@ PROVIDER_WORKFLOWS = [
     ),
     ProviderWorkflow(
         ingester_class=WikimediaCommonsDataIngester,
-        start_date=datetime(2020, 11, 1),
+        start_date=datetime(2023, 2, 1),
         schedule_string="@daily",
         dated=True,
         pull_timeout=timedelta(hours=12),
```


4. Trigger Wikimedia DAG with config `{"date": "2023-02-10"}`
5. This kicks off a scheduled DAG because the DAG hasn't run yet, go ahead and mark the scheduled run as failed

![image](https://user-images.githubusercontent.com/10214785/220521088-b236008d-e6a6-4867-b327-5b7bc3810ddb.png)


6. Wait for the run to complete (~40 minutes, if it _does_ complete that's evidence that this works because this date hit the timeout!)
7. Run `just db-shell` then execute `select * from image where provider='wikimedia' and md5(foreign_identifier)=md5('4298234');`, this will return the result for the [wikinews commentary icon](https://commons.wikimedia.org/w/index.php?curid=4298234) which was returning thousands of global usage results.
8. Confirm that `meta_data` field for the result above has a `global_usage_count` that is about 2k
9. Confirm in the logs for the task that it says "Hit iteration count limit for <some token>"

## Checklist



      `Update index.md`).
      a parent feature branch.
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin



<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
